drivers/bluetooth/hci/hci_b91.c

```diff
@@ -0,0 +1,303 @@
+	data += sizeof(hdr);
+	len -= sizeof(hdr);
+
+	if (len != hdr.len) {
+		BT_ERR("Event payload length is not correct");
+		return NULL;
+	}
+	BT_DBG("len %u", hdr.len);
+
+	buf = bt_buf_get_evt(hdr.evt, discardable, K_NO_WAIT);
+	if (!buf) {
+		if (discardable) {
+			BT_DBG("Discardable buffer pool full, ignoring event");
+		} else {
+			BT_ERR("No available event buffers!");
+		}
+		return buf;
+	}
+
+	net_buf_add_mem(buf, &hdr, sizeof(hdr));
+
+	buf_tailroom = net_buf_tailroom(buf);
+	if (buf_tailroom < len) {
+		BT_ERR("Not enough space in buffer %zu/%zu",
+		       len, buf_tailroom);
+		net_buf_unref(buf);
+		return NULL;
+	}
+
+	net_buf_add_mem(buf, data, len);
+
+	return buf;
+}
+
+static struct net_buf *bt_b91_acl_recv(uint8_t *data, size_t len)
+{
+	struct bt_hci_acl_hdr hdr;
+	struct net_buf *buf;
+	size_t buf_tailroom;
+
+	if (len < sizeof(hdr)) {
+		BT_ERR("Not enough data for ACL header");
+		return NULL;
+	}
+
+	buf = bt_buf_get_rx(BT_BUF_ACL_IN, K_NO_WAIT);
+	if (buf) {
+		memcpy((void *)&hdr, data, sizeof(hdr));
+		data += sizeof(hdr);
+		len -= sizeof(hdr);
+	} else {
+		BT_ERR("No available ACL buffers!");
+		return NULL;
+	}
+
+	if (len != sys_le16_to_cpu(hdr.len)) {
+		BT_ERR("ACL payload length is not correct");
+		net_buf_unref(buf);
+		return NULL;
+	}
+
+	net_buf_add_mem(buf, &hdr, sizeof(hdr));
+	buf_tailroom = net_buf_tailroom(buf);
+	if (buf_tailroom < len) {
+		BT_ERR("Not enough space in buffer %zu/%zu",
+		       len, buf_tailroom);
+		net_buf_unref(buf);
+		return NULL;
+	}
+
+	BT_DBG("len %u", len);
+	net_buf_add_mem(buf, data, len);
+
+	return buf;
+}
+
+static struct net_buf *bt_b91_iso_recv(uint8_t *data, size_t len)
+{
+	struct bt_hci_iso_hdr hdr;
+	struct net_buf *buf;
+	size_t buf_tailroom;
+
+	if (len < sizeof(hdr)) {
+		BT_ERR("Not enough data for ISO header");
+		return NULL;
+	}
+
+	buf = bt_buf_get_rx(BT_BUF_ISO_IN, K_NO_WAIT);
+	if (buf) {
+		memcpy((void *)&hdr, data, sizeof(hdr));
+		data += sizeof(hdr);
+		len -= sizeof(hdr);
+
+		net_buf_add_mem(buf, &hdr, sizeof(hdr));
+	} else {
+		BT_ERR("No available ISO buffers!");
+		return NULL;
+	}
+
+	if (len != bt_iso_hdr_len(sys_le16_to_cpu(hdr.len))) {
+		BT_ERR("ISO payload length is not correct");
+		net_buf_unref(buf);
+		return NULL;
+	}
+
+	buf_tailroom = net_buf_tailroom(buf);
+	if (buf_tailroom < len) {
+		BT_ERR("Not enough space in buffer %zu/%zu",
+		       len, buf_tailroom);
+		net_buf_unref(buf);
+		return NULL;
+	}
+
+	BT_DBG("len %zu", len);
+	net_buf_add_mem(buf, data, len);
+
+	return buf;
+}
+
+static void hci_b91_host_rcv_pkt(uint8_t *data, uint16_t len)
+{
+	uint8_t pkt_indicator;
+	struct net_buf *buf;
+
+	BT_HEXDUMP_DBG(data, len, "host packet data:");
+
+	pkt_indicator = *data++;
+	len -= sizeof(pkt_indicator);
+
+	switch (pkt_indicator) {
+	case HCI_EVT:
+		buf = bt_b91_evt_recv(data, len);
+		break;
+
+	case HCI_ACL:
+		buf = bt_b91_acl_recv(data, len);
+		break;
+
+	case HCI_SCO:
+		buf = bt_b91_iso_recv(data, len);
```

Comment:
Could you explain this to me. SCO is a BR/EDR feature, but the function you're calling is called ISO. Does this controller support either of these features?

Reply:
Fixed,
Added BT_HCI_TX_STACK_SIZE and BT_HCI_RESERVE default values.
Also removed all SCO/ISO related code (the controller supports them, however currently we are not contributing them).